workshop/stringBasic/StringBasic.cpp

```diff
@@ -0,0 +1,245 @@
+    if (m_string != nullptr) {
+        try {
+            delete[] m_string;
+        }
+        catch (...) {
+        }
+    }
+
+    *this = StringBasic(other);
+    return *this;
+}
+
+
+StringBasic::StringBasic(StringBasic&& other) noexcept :
+    m_size(other.m_size),
+    m_lenght(other.m_lenght),
+    m_string(other.m_string)
+{
+    other.m_string = nullptr;
+    other.m_lenght = 0;
+    other.m_size = 0;
+}
+
+
+StringBasic& StringBasic::operator=(StringBasic&& other) noexcept {
+    if (m_string != nullptr) {
+        try {
+            delete[] m_string;
+        }
+        catch (...) {
+        }
+    }
+
+    other.m_string = nullptr;
+    *this = std::move(other);
+    return *this;
+}
+
+
+const unsigned long StringBasic::getLenght() const noexcept {
+    return m_lenght;
+}
+
+
+const char StringBasic::operator[](std::size_t index) const{
+    if (index >= m_lenght) {
+        throw std::out_of_range("invalid index exception");
+    }
+
+    return m_string[index];
+}
+
+
+char& StringBasic::operator[](std::size_t index) {
+    if (index >= m_lenght) {
+        throw std::out_of_range("invalid index exception");
+    }
+
+    return m_string[index];
+}
+
+
+std::ostream& operator<<(std::ostream& strm, const StringBasic& str) noexcept {
+    for (std::size_t i = 0; i < str.m_lenght; i++) {
+        strm << str.m_string[i];
+    }
+    return strm;
+}
+
+
+StringBasic& StringBasic::operator+=(const StringBasic& rhs) {
+    add_string(rhs.m_string);
+    return *this;
+}
+
+
+StringBasic& StringBasic::operator+=(const char& rhs) {
+    add_char(rhs);
+    return *this;
+}
+
+
+StringBasic& StringBasic::operator+=(const char* const rhs) {
+    add_string(rhs);
+    return *this;
+}
+
+
+StringBasic StringBasic::operator+(const StringBasic& rhs) {
+    StringBasic string = *this;
+    string.add_string(rhs.m_string);
+    return string;
+}
+
+
+StringBasic StringBasic::operator+(const char* const rhs) {
+    StringBasic string = *this;
+    string.add_string(rhs);
+    return string;
+}
+
+
+StringBasic StringBasic::operator+(const char& rhs) {
+    StringBasic string = *this;
+    string.add_char(rhs);
+    return string;
+}
+
+
+void StringBasic::add_char(const char chr) {
+    if (m_size == 0) {
+        m_size = 2;
+        m_string = new char[m_size];
+        std::memset(m_string, 0, m_size);
+        m_string[m_lenght] = chr;
+        m_lenght++;;
+    }
+    else if (m_lenght == (m_size - 1)) {
+        increase_size(SIZE_SCALE * m_size);
+    }
+    else {
+        m_string[m_lenght] = chr;
+        m_lenght++;
+    }
+}
+
+
+void StringBasic::add_string(const char* const str) {
+    for (std::size_t i = 0; str[i]; i++) {
+        add_char(str[i]);
+    }
+}
+
+
+void StringBasic::insert(StringBasic& str, std::size_t offset) {
+    if (offset > m_lenght) {
+        throw std::out_of_range("invalid offset");
+    }
+    if ((m_lenght + str.m_lenght) > (m_size - 1)) {
+        increase_size(SIZE_SCALE + str.m_lenght);
```

Comment:
Why do we need the SIZE_SCALE const?
This seems like a weird premature optimization.
And you know what they say: "Premature optimization is the root of all evil".